File: experiment-4/runs/language=clojure_model=claude-opus-4-8_tooling=none/rep1/src/brazilian_soccer/mcp.clj
(ns brazilian-soccer.mcp
  "=============================================================================
   Brazilian Soccer MCP Server - MCP / JSON-RPC Protocol Layer
   =============================================================================

   CONTEXT
     Implements the Model Context Protocol over the stdio transport using
     newline-delimited JSON-RPC 2.0 messages. This is the boundary an LLM host
     (Claude Desktop, etc.) talks to. It exposes the soccer knowledge graph as
     a set of callable MCP 'tools', each backed by a pure function in
     brazilian-soccer.query and rendered by brazilian-soccer.format.

     Supported JSON-RPC methods:
       initialize                 - capability handshake
       notifications/initialized  - client ready (no response)
       tools/list                 - advertise available tools + JSON schemas
       tools/call                 - invoke a tool by name with arguments
       ping                       - liveness check

   DESIGN
     `handle-request` is a pure function (db + request-map -> response-map|nil)
     so the entire protocol surface is unit-testable without any IO. `serve!`
     wires it to real stdin/stdout. Tool handlers receive (db args) and return
     a human-readable string; `tools/call` wraps that into MCP content.

   PUBLIC API
     tools           - vector of tool definitions
     handle-request  - pure JSON-RPC request dispatch
     serve!          - run the stdio read/dispatch/write loop
   ============================================================================="
  (:require [cheshire.core :as json]
            [clojure.string :as str]
            [brazilian-soccer.query :as q]
            [brazilian-soccer.format :as fmt]))

(def protocol-version "2024-11-05")

(def server-info
  {:name "brazilian-soccer-mcp" :version "1.0.0"})

;; ----------------------------------------------------------------------------
;; Argument coercion helpers (JSON args arrive as strings/numbers)
;; ----------------------------------------------------------------------------

(defn- ->int [v]
  (cond (number? v) (long v)
        (string? v) (try (Long/parseLong (str/trim v)) (catch Exception _ nil))
        :else nil))

(defn- ->kw-side [v]
  (case (some-> v str str/lower-case)
    "home" :home "away" :away :either))

(defn- non-blank [v] (when (and v (not (str/blank? (str v)))) (str v)))

;; ----------------------------------------------------------------------------
;; Tool definitions
;;   Each: {:name :description :inputSchema :handler}
;;   :handler is (fn [db args] -> string)
;; ----------------------------------------------------------------------------

(defn- prop [type desc & {:as extra}]
  (merge {:type type :description desc} extra))

(def tools
  [{:name "search_matches"
    :description "Find soccer matches by team, opponent, competition, season, or date range. Returns a readable list of matches with scores and competition context."
    :inputSchema {:type "object"
                  :properties {:team        (prop "string" "Team name (e.g. \"Flamengo\"). Matches home, away, or either side.")
                               :side        (prop "string" "Which side the team must be on: \"home\", \"away\", or \"either\" (default).")
                               :opponent    (prop "string" "Second team; both teams must appear in the match.")
                               :competition (prop "string" "Competition filter, e.g. \"Brasileirão\", \"Copa do Brasil\", \"Libertadores\".")
                               :season      (prop "integer" "Season year, e.g. 2019.")
                               :date_from   (prop "string" "Inclusive ISO start date YYYY-MM-DD.")
                               :date_to     (prop "string" "Inclusive ISO end date YYYY-MM-DD.")
                               :limit       (prop "integer" "Max matches to return (default 20).")}}
    :handler (fn [db a]
               (let [limit (or (->int (:limit a)) 20)
                     opts {:team (non-blank (:team a))
                           :side (->kw-side (:side a))
                           :opponent (non-blank (:opponent a))
                           :competition (non-blank (:competition a))
                           :season (->int (:season a))
                           :date-from (non-blank (:date_from a))
                           :date-to (non-blank (:date_to a))
                           :limit limit}
                     ;; Get the true total then the capped page for the "N more" note.
                     total (count (q/search-matches db (assoc opts :limit Long/MAX_VALUE)))
                     page (q/search-matches db opts)]
                 (str (if (:opponent a)
                        (format "Matches between %s and %s:\n" (:team a) (:opponent a))
                        "")
                      (fmt/matches page total))))}

   {:name "team_stats"
    :description "Win/loss/draw record, goals for/against, and win rate for a team, optionally filtered by season, competition, and venue (home/away)."
    :inputSchema {:type "object"
                  :properties {:team        (prop "string" "Team name (required).")
                               :season      (prop "integer" "Season year filter.")
                               :competition (prop "string" "Competition filter.")
                               :venue       (prop "string" "\"home\", \"away\", or \"all\" (default).")}
                  :required ["team"]}
    :handler (fn [db a]
               (fmt/team-stats
                (q/team-stats db (:team a)
                              {:season (->int (:season a))
                               :competition (non-blank (:competition a))
                               :venue (case (some-> (:venue a) str str/lower-case)
                                        "home" :home "away" :away :all)})))}

   {:name "head_to_head"
    :description "Head-to-head record between two teams across all competitions, with recent matches."
    :inputSchema {:type "object"
                  :properties {:team1 (prop "string" "First team (required).")
                               :team2 (prop "string" "Second team (required).")
                               :show  (prop "integer" "How many recent matches to list (default 5).")}
                  :required ["team1" "team2"]}
    :handler (fn [db a]
               (fmt/head-to-head (q/head-to-head db (:team1 a) (:team2 a))
                                 (or (->int (:show a)) 5)))}

   {:name "search_players"
    :description "Search FIFA player data by name, nationality, club, position, or minimum overall rating. Sorted by rating."
    :inputSchema {:type "object"
                  :properties {:name        (prop "string" "Player name substring, e.g. \"Neymar\".")
                               :nationality (prop "string" "Nationality, e.g. \"Brazil\".")
                               :club        (prop "string" "Club name substring.")
                               :position    (prop "string" "Position code, e.g. \"ST\", \"GK\", \"LW\".")
                               :min_overall (prop "integer" "Minimum FIFA overall rating.")
                               :limit       (prop "integer" "Max players to return (default 10).")}}
    :handler (fn [db a]
               (fmt/players
                (q/search-players db {:name (non-blank (:name a))
                                      :nationality (non-blank (:nationality a))
                                      :club (non-blank (:club a))
                                      :position (non-blank (:position a))
                                      :min-overall (->int (:min_overall a))
                                      :limit (or (->int (:limit a)) 10)})))}

   {:name "club_roster"
    :description "List players belonging to a club with the squad size and average overall rating."
    :inputSchema {:type "object"
                  :properties {:club (prop "string" "Club name (required), e.g. \"Santos\".")}
                  :required ["club"]}
    :handler (fn [db a] (fmt/club-roster (q/club-roster db (:club a))))}

   {:name "standings"
    :description "Computed league standings (3pts win, 1pt draw) for a competition and season, derived from match results."
    :inputSchema {:type "object"
                  :properties {:competition (prop "string" "Competition, e.g. \"Brasileirão Série A\".")
                               :season      (prop "integer" "Season year (required).")
                               :top         (prop "integer" "Show only the top N rows (default all).")}
                  :required ["season"]}
    :handler (fn [db a]
               (let [season (->int (:season a))
                     comp (or (non-blank (:competition a)) "Brasileirão Série A")
                     rows (q/standings db {:competition comp :season season})]
                 (str (format "%s %d Final Standings (calculated from matches):\n" comp season)
                      (fmt/standings rows (or (->int (:top a)) 0)))))}

   {:name "champion"
    :description "The calculated champion (top of the standings) for a competition and season."
    :inputSchema {:type "object"
                  :properties {:competition (prop "string" "Competition (default \"Brasileirão Série A\").")
                               :season      (prop "integer" "Season year (required).")}
                  :required ["season"]}
    :handler (fn [db a]
               (let [season (->int (:season a))
                     comp (or (non-blank (:competition a)) "Brasileirão Série A")
                     champ (q/champion db {:competition comp :season season})]
                 (if champ
                   (format "%s %d champion (calculated): %s with %d points (%dW, %dD, %dL)."
                           comp season (:team champ) (:points champ)
                           (:win champ) (:draw champ) (:loss champ))
                   (format "No data for %s %d." comp season))))}

   {:name "competition_stats"
    :description "Aggregate statistics for a competition/season: match count, total & average goals, home/away win rates."
    :inputSchema {:type "object"
                  :properties {:competition (prop "string" "Competition filter.")
                               :season      (prop "integer" "Season year filter.")}}
    :handler (fn [db a]
               (fmt/competition-stats
                (q/competition-stats db {:competition (non-blank (:competition a))
                                         :season (->int (:season a))})))}

   {:name "biggest_wins"
    :description "Matches with the largest goal margin, optionally filtered by competition/season."
    :inputSchema {:type "object"
                  :properties {:competition (prop "string" "Competition filter.")
                               :season      (prop "integer" "Season year filter.")
                               :limit       (prop "integer" "How many matches (default 10).")}}
    :handler (fn [db a]
               (let [ms (q/biggest-wins db {:competition (non-blank (:competition a))
                                            :season (->int (:season a))
                                            :limit (or (->int (:limit a)) 10)})]
                 (str "Biggest victories:\n" (fmt/matches ms))))}

   {:name "best_record"
    :description "Teams ranked by win rate within a competition/season (home/away/all), with a minimum games threshold."
    :inputSchema {:type "object"
                  :properties {:competition (prop "string" "Competition filter.")
                               :season      (prop "integer" "Season year filter.")
                               :venue       (prop "string" "\"home\", \"away\", or \"all\" (default).")
                               :limit       (prop "integer" "How many teams (default 10).")}}
    :handler (fn [db a]
               (let [rows (q/best-record db {:competition (non-blank (:competition a))
                                             :season (->int (:season a))
                                             :venue (case (some-> (:venue a) str str/lower-case)
                                                      "home" :home "away" :away :all)
                                             :limit (or (->int (:limit a)) 10)})]
                 (if (empty? rows)
                   "No teams matched the criteria."
                   (str/join "\n"
                             (map-indexed
                              (fn [i s]
                                (format "%d. %s - %.1f%% win rate (%dW, %dD, %dL)"
                                        (inc i) (:team s) (:win-rate s)
                                        (:win s) (:draw s) (:loss s)))
                              rows)))))}

   {:name "list_competitions"
    :description "List all competitions available in the dataset."
    :inputSchema {:type "object" :properties {}}
    :handler (fn [db _] (str "Competitions:\n- " (str/join "\n- " (q/list-competitions db))))}

   {:name "list_seasons"
    :description "List the seasons available, optionally for one competition."
    :inputSchema {:type "object"
                  :properties {:competition (prop "string" "Competition filter (optional).")}}
    :handler (fn [db a]
               (str "Seasons: " (str/join ", " (q/list-seasons db (non-blank (:competition a))))))}])

(def ^:private tool-index
  (into {} (map (juxt :name identity) tools)))

;; ----------------------------------------------------------------------------
;; JSON-RPC plumbing
;; ----------------------------------------------------------------------------

(defn- result [id v] {:jsonrpc "2.0" :id id :result v})
(defn- error  [id code msg] {:jsonrpc "2.0" :id id :error {:code code :message msg}})

(defn- tool-listing
  "Public schema view of the tools (drops the internal :handler)."
  []
  (mapv #(select-keys % [:name :description :inputSchema]) tools))

(defn- call-tool
  "Invoke a named tool. Returns an MCP tools/call result map (with :isError on
   failure) - never throws."
  [db name args]
  (if-let [tool (get tool-index name)]
    (try
      (let [text ((:handler tool) db (or args {}))]
        {:content [{:type "text" :text text}] :isError false})
      (catch Exception e
        {:content [{:type "text" :text (str "Error running tool '" name "': " (.getMessage e))}]
         :isError true}))
    {:content [{:type "text" :text (str "Unknown tool: " name)}] :isError true}))

(defn handle-request
  "Pure dispatch for a single parsed JSON-RPC request map. Returns the response
   map to send back, or nil for notifications (which take no response).
   `db` is the loaded knowledge graph."
  [db {:keys [id method params]}]
  (case method
    "initialize"
    (result id {:protocolVersion protocol-version
                :capabilities {:tools {}}
                :serverInfo server-info})

    "notifications/initialized" nil
    "initialized" nil

    "ping" (result id {})

    "tools/list" (result id {:tools (tool-listing)})

    "tools/call"
    (let [{:keys [name arguments]} params]
      (result id (call-tool db name arguments)))

    ;; Unknown method: only error on requests (those with an id), ignore notifications.
    (if (nil? id)
      nil
      (error id -32601 (str "Method not found: " method)))))

;; ----------------------------------------------------------------------------
;; stdio transport
;; ----------------------------------------------------------------------------

(defn- parse-request [line]
  (try (json/parse-string line true)
       (catch Exception _ ::parse-error)))

(defn serve!
  "Run the blocking stdio JSON-RPC loop: read newline-delimited JSON requests
   from `in` (default *in*), dispatch against `db`, and write newline-delimited
   JSON responses to `out` (default *out*). Returns when EOF is reached."
  ([db] (serve! db *in* *out*))
  ([db in out]
   (let [reader (clojure.java.io/reader in)
         writer (clojure.java.io/writer out)
         emit (fn [resp]
                (when resp
                  (.write writer (json/generate-string resp))
                  (.write writer "\n")
                  (.flush writer)))]
     (loop []
       (when-let [line (.readLine reader)]
         (when-not (str/blank? line)
           (let [req (parse-request line)]
             (if (= req ::parse-error)
               (emit (error nil -32700 "Parse error"))
               (emit (handle-request db req)))))
         (recur))))))

File: experiments/adrianco/experiment-5/runs/language=rust_model=claude-opus-4-8_tooling=none/rep1/src/mcp.rs
//! Minimal MCP (Model Context Protocol) server over a newline-delimited
//! JSON-RPC 2.0 stdio transport. Implements `initialize`, `tools/list`,
//! `tools/call` and `ping`.

use std::io::{BufRead, Write};

use serde_json::{json, Value};

use crate::data::Dataset;
use crate::queries::{self, MatchQuery, PlayerQuery, Venue};

pub const PROTOCOL_VERSION: &str = "2024-11-05";

/// Run the stdio server loop, reading requests from `reader` and writing
/// responses to `writer`. Each message is a single line of JSON.
pub fn serve(ds: &Dataset, reader: impl BufRead, mut writer: impl Write) -> std::io::Result<()> {
    for line in reader.lines() {
        let line = line?;
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let request: Value = match serde_json::from_str(trimmed) {
            Ok(v) => v,
            Err(e) => {
                let resp = error_response(Value::Null, -32700, &format!("Parse error: {e}"));
                writeln!(writer, "{resp}")?;
                writer.flush()?;
                continue;
            }
        };

        if let Some(resp) = handle_request(ds, &request) {
            writeln!(writer, "{resp}")?;
            writer.flush()?;
        }
    }
    Ok(())
}

/// Handle one JSON-RPC request. Returns `None` for notifications (no `id`),
/// which must not produce a response.
pub fn handle_request(ds: &Dataset, req: &Value) -> Option<Value> {
    let method = req.get("method").and_then(Value::as_str).unwrap_or("");
    let id = req.get("id").cloned();
    let is_notification = id.is_none();

    match method {
        "initialize" => Some(success(
            id.unwrap_or(Value::Null),
            json!({
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": { "tools": {} },
                "serverInfo": {
                    "name": "brazilian-soccer-mcp",
                    "version": env!("CARGO_PKG_VERSION"),
                },
            }),
        )),
        "notifications/initialized" | "initialized" => None,
        "ping" => Some(success(id.unwrap_or(Value::Null), json!({}))),
        "tools/list" => Some(success(id.unwrap_or(Value::Null), json!({ "tools": tool_definitions() }))),
        "tools/call" => {
            let id = id.unwrap_or(Value::Null);
            let params = req.get("params").cloned().unwrap_or(Value::Null);
            let name = params.get("name").and_then(Value::as_str).unwrap_or("");
            let args = params.get("arguments").cloned().unwrap_or_else(|| json!({}));
            match dispatch_tool(ds, name, &args) {
                Ok(text) => Some(success(
                    id,
                    json!({ "content": [{ "type": "text", "text": text }] }),
                )),
                Err(msg) => Some(success(
                    id,
                    json!({
                        "content": [{ "type": "text", "text": msg }],
                        "isError": true,
                    }),
                )),
            }
        }
        _ => {
            if is_notification {
                None
            } else {
                Some(error_response(
                    id.unwrap_or(Value::Null),
                    -32601,
                    &format!("Method not found: {method}"),
                ))
            }
        }
    }
}

fn success(id: Value, result: Value) -> Value {
    json!({ "jsonrpc": "2.0", "id": id, "result": result })
}

fn error_response(id: Value, code: i64, message: &str) -> Value {
    json!({ "jsonrpc": "2.0", "id": id, "error": { "code": code, "message": message } })
}

// ---------------------------------------------------------------------------
// Argument helpers
// ---------------------------------------------------------------------------

fn arg_str<'a>(args: &'a Value, key: &str) -> Option<&'a str> {
    args.get(key).and_then(Value::as_str).filter(|s| !s.trim().is_empty())
}

fn arg_i32(args: &Value, key: &str) -> Option<i32> {
    match args.get(key) {
        Some(Value::Number(n)) => n.as_i64().map(|v| v as i32),
        Some(Value::String(s)) => s.trim().parse::<i32>().ok(),
        _ => None,
    }
}

fn arg_usize(args: &Value, key: &str) -> usize {
    match args.get(key) {
        Some(Value::Number(n)) => n.as_u64().unwrap_or(0) as usize,
        Some(Value::String(s)) => s.trim().parse::<usize>().unwrap_or(0),
        _ => 0,
    }
}

// ---------------------------------------------------------------------------
// Tool dispatch
// ---------------------------------------------------------------------------

/// Execute a named tool. Public so it can be exercised directly in tests and
/// from the CLI without going through JSON-RPC framing.
pub fn dispatch_tool(ds: &Dataset, name: &str, args: &Value) -> Result<String, String> {
    match name {
        "search_matches" => Ok(queries::search_matches(
            ds,
            &MatchQuery {
                team: arg_str(args, "team"),
                team2: arg_str(args, "team2"),
                competition: arg_str(args, "competition"),
                season: arg_i32(args, "season"),
                date_from: arg_str(args, "date_from"),
                date_to: arg_str(args, "date_to"),
                limit: arg_usize(args, "limit"),
            },
        )),
        "head_to_head" => {
            let t1 = arg_str(args, "team1").ok_or("`team1` is required")?;
            let t2 = arg_str(args, "team2").ok_or("`team2` is required")?;
            Ok(queries::head_to_head(
                ds,
                t1,
                t2,
                arg_str(args, "competition"),
                arg_i32(args, "season"),
            ))
        }
        "team_stats" => {
            let team = arg_str(args, "team").ok_or("`team` is required")?;
            let venue = match arg_str(args, "venue").map(|s| s.to_lowercase()) {
                Some(ref v) if v == "home" => Venue::Home,
                Some(ref v) if v == "away" => Venue::Away,
                _ => Venue::All,
            };
            Ok(queries::team_stats(
                ds,
                team,
                arg_i32(args, "season"),
                arg_str(args, "competition"),
                venue,
            ))
        }
        "standings" => {
            let season = arg_i32(args, "season").ok_or("`season` is required")?;
            Ok(queries::standings(ds, season, arg_str(args, "competition")))
        }
        "competition_stats" => Ok(queries::competition_stats(
            ds,
            arg_str(args, "competition"),
            arg_i32(args, "season"),
        )),
        "search_players" => Ok(queries::search_players(
            ds,
            &PlayerQuery {
                name: arg_str(args, "name"),
                nationality: arg_str(args, "nationality"),
                club: arg_str(args, "club"),
                position: arg_str(args, "position"),
                min_overall: arg_i32(args, "min_overall"),
                limit: arg_usize(args, "limit"),
            },
        )),
        "top_players" => Ok(queries::top_players(
            ds,
            arg_str(args, "nationality"),
            arg_str(args, "club"),
            arg_str(args, "position"),
            arg_usize(args, "limit"),
        )),
        "list_datasets" => Ok(queries::list_datasets(ds)),
        other => Err(format!("Unknown tool: {other}")),
    }
}

// ---------------------------------------------------------------------------
// Tool schema definitions
// ---------------------------------------------------------------------------

fn str_prop(desc: &str) -> Value {
    json!({ "type": "string", "description": desc })
}
fn int_prop(desc: &str) -> Value {
    json!({ "type": "integer", "description": desc })
}

pub fn tool_definitions() -> Vec<Value> {
    vec![
        json!({
            "name": "search_matches",
            "description": "Search matches by team, opponent, competition, season and/or date range. \
                When two teams are given, includes a head-to-head summary. Competitions: Brasileirão, \
                Copa do Brasil, Copa Libertadores.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team": str_prop("Team name (matches home or away). Variants like 'Flamengo' or 'Flamengo-RJ' work."),
                    "team2": str_prop("Optional second team to find matches between the two."),
                    "competition": str_prop("Competition filter, e.g. 'Brasileirão', 'Copa do Brasil', 'Libertadores'."),
                    "season": int_prop("Season year, e.g. 2019."),
                    "date_from": str_prop("Earliest match date, ISO YYYY-MM-DD."),
                    "date_to": str_prop("Latest match date, ISO YYYY-MM-DD."),
                    "limit": int_prop("Max matches to list (default 25)."),
                },
            },
        }),
        json!({
            "name": "head_to_head",
            "description": "Head-to-head record and recent matches between two teams.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team1": str_prop("First team."),
                    "team2": str_prop("Second team."),
                    "competition": str_prop("Optional competition filter."),
                    "season": int_prop("Optional season filter."),
                },
                "required": ["team1", "team2"],
            },
        }),
        json!({
            "name": "team_stats",
            "description": "Win/draw/loss record, goals for/against and win rate for a team, \
                optionally filtered by season, competition and venue (home/away/all).",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "team": str_prop("Team name."),
                    "season": int_prop("Optional season year."),
                    "competition": str_prop("Optional competition filter."),
                    "venue": json!({ "type": "string", "enum": ["all", "home", "away"], "description": "Restrict to home or away matches (default all)." }),
                },
                "required": ["team"],
            },
        }),
        json!({
            "name": "standings",
            "description": "Final league standings for a season, calculated from match results \
                (3 points per win, 1 per draw). Defaults to the Brasileirão.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "season": int_prop("Season year, e.g. 2019."),
                    "competition": str_prop("Competition (default 'Brasileirão')."),
                },
                "required": ["season"],
            },
        }),
        json!({
            "name": "competition_stats",
            "description": "Aggregate statistics: average goals per match, home/draw/away win rates, \
                biggest victories and top scoring teams. Filter by competition and/or season.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "competition": str_prop("Optional competition filter."),
                    "season": int_prop("Optional season year."),
                },
            },
        }),
        json!({
            "name": "search_players",
            "description": "Search FIFA player data by name, nationality, club, position and minimum \
                overall rating. Results are sorted by overall rating (highest first).",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "name": str_prop("Player name substring, e.g. 'Neymar'."),
                    "nationality": str_prop("Exact nationality, e.g. 'Brazil'."),
                    "club": str_prop("Club name substring, e.g. 'Flamengo'."),
                    "position": str_prop("Exact position code, e.g. 'GK', 'ST', 'CB'."),
                    "min_overall": int_prop("Minimum FIFA overall rating."),
                    "limit": int_prop("Max players to list (default 20)."),
                },
            },
        }),
        json!({
            "name": "top_players",
            "description": "Highest-rated players for a given nationality, club and/or position. \
                When a nationality is supplied, also returns a per-club breakdown with average ratings.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "nationality": str_prop("Nationality, e.g. 'Brazil'."),
                    "club": str_prop("Club name substring."),
                    "position": str_prop("Position code."),
                    "limit": int_prop("Max players to list (default 20)."),
                },
            },
        }),
        json!({
            "name": "list_datasets",
            "description": "Summary of loaded datasets and row counts (data coverage).",
            "inputSchema": { "type": "object", "properties": {} },
        }),
    ]
}
